test/junit/scala/tools/nsc/backend/jvm/NestedClassesCollectorTest.scala

```diff
@@ -0,0 +1,115 @@
+    ref("Lp/Kl.Ne.In;", List("p/Kl", "p/Kl$Ne", "p/Kl$Ne$In"))
+    ref("LA<*>;", List("A"))
+    ref("LA<**+[I[JTFoo;-TBar;LB;*>;", List("A", "B"))
+    ref("Lp/A<[I[LTBoo<*>;-[JTFoo;-TBar;Lp/B<[J+[Lp/C;>.N<+TT;*Lp/D;>;*>;", List("TBoo", "p/A", "p/B", "p/B$N", "p/C", "p/D"))
+    ref("Lp/A<[I[Lp/B<*>;>;", List("p/A", "p/B"))
+    ref("Lp/A<Lp/B;>.C<Lp/D;>.E;", List("p/A", "p/A$C", "p/A$C$E", "p/B", "p/D"))
+
+    ref("[I", Nil)
+    ref("[[[LA;", List("A"))
+    ref("[[[LA<**+[I-[LB;>;", List("A", "B"))
+  }
+
+  @Test
+  def classSignatures(): Unit = {
+    def cls(sig: String, expect: List[String]) = {
+      c.visitClassSignature(sig)
+      assertEquals(inners, expect)
+    }
+
+    cls("LA;", List("A"))
+    cls("LA;LB;", List("A", "B"))
+    cls("Lp/a/A;Lp/B;", List("p/B", "p/a/A"))
+    cls("<T:>LA;", List("A"))
+    cls("<T:LA;:[I:TU;:[TV;>LB;", List("A", "B"))
+    cls("<T:LA;:[Lp/B<+Lp/C;>;:TU;:[TV;>LC;", List("A", "C", "p/B", "p/C"))
+    cls("<T::TT;>LA;", List("A"))
+    cls("<T:LA;>LB;", List("A", "B")) // one type parameter T with class bound A
+    cls("<P:TT;>LA;", List("A")) // one type parameter
+
+    // Missing ClassBound without an interface bound. Probably the grammar only allows those by
+    // accident. Our parser doesn't. https://stackoverflow.com/q/44284928
+    // cls("<T:U:>LA;", List("A"))
+    // cls("<T:L:>LA;", List("A"))
+    // cls("<T:LA:>LB;", List("B")) // two type parameters, T and LA
+    // cls("<P:TT:>LA;", List("A")) // two type parameters
+    // cls("<T:U:LA;>LB;", List("A", "B"))
+  }
+
+  @Test
+  def methodSignatures(): Unit = {
+    def met(sig: String, expect: List[String]) = {
+      c.visitMethodSignature(sig)
+      assertEquals(inners, expect)
+    }
+
+    // type parameters implementation is the same as for class signatures, so only basic testing here
+    met("()V", Nil)
+    met("(BJI)Z", Nil)
+    met("(IJLp/A;Z)Lp/B;", List("p/A", "p/B"))
+    met("<X:LA;:[I:TU;:[TV;Y:[I:LB<+LC;>;>([I[[[LD<**>;)TT;", List("A", "B", "C", "D"))
+    met("(LA;ITT;)I^LB;", List("A", "B"))
+    met("()I^TT;^Lp/A<**+[[Lp/B;>;^TBA;", List("p/A", "p/B"))
+    met("()V^TT;", Nil)
+  }
+
+  @Test
+  @Ignore("manually run test")
+  def rtJar(): Unit = {
+    import java.nio.file._
+    import scala.collection.JavaConverters._
+    val zipfile = Paths.get("/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/rt.jar")
+    val fs = FileSystems.newFileSystem(zipfile, null)
+    val root = fs.getRootDirectories.iterator().next()
+    val contents = Files.walk(root).iterator().asScala.toList
+    for (f <- contents if Files.isRegularFile(f) && f.getFileName.toString.endsWith(".class")) {
+      val classNode = AsmUtils.classFromBytes(Files.readAllBytes(f))
+      c.visitClassSignature(classNode.signature)
+      classNode.methods.iterator().asScala.map(_.signature).foreach(c.visitMethodSignature)
+      classNode.fields.iterator().asScala.map(_.signature).foreach(c.visitFieldSignature)
+    }
+  }
```

Comment:
I ran this across the contents of my Ivy cache.

## Invalid signature

Needs investigation.

```
/Users/jz/.ivy2/cache/org.scala-sbt/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0/jars/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0-0.13.13_20161028T115300.jar

currentClass = /xsbt/Message.class
java.lang.Exception: Out of bounds, 63 >= 63 <T:Ljava/lang/Object;()>(Lscala/Function0<TT;>;)Lxsbti/F0<TT;>;
	at scala.tools.nsc.backend.jvm.Collector.$anonfun$raiseError$1(NestedClassesCollectorTest.scala:19)
	at scala.Option.getOrElse(Option.scala:121)
	at scala.tools.nsc.backend.jvm.Collector.raiseError(NestedClassesCollectorTest.scala:19)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.current(BackendUtils.scala:613)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.skipUntil(BackendUtils.scala:631)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.typeParameters(BackendUtils.scala:687)

```

```
% javap -cp /Users/jz/.ivy2/cache/org.scala-sbt/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0/jars/org.scala-sbt-compiler-interface-0.13.13-bin_2.12.2-20170328-073517-729534b__52.0-0.13.13_20161028T115300.jar xsbt/Message
Compiled from "Message.scala"
public final class xsbt.Message {
Error: A serious internal error has occurred: java.lang.StringIndexOutOfBoundsException: String index out of range: -22
Please file a bug report, and include the following information:
java.lang.StringIndexOutOfBoundsException: String index out of range: -22
	at java.lang.String.substring(String.java:1967)
```

## Invalid signature due to value class wrapping

As fixed in this PR.

```
currentClass = /org/scalacheck/util/Buildable$$anon$5.class
/Users/jz/.ivy2/cache/org.scala-tools.testing/scalacheck_2.9.0-1/jars/scalacheck_2.9.0-1-1.9.jar	
scala.MatchError: > (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.typeArguments(BackendUtils.scala:653)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:664)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.classSignature(BackendUtils.scala:699)
cu
```

```
currentClass = /org/specs2/internal/scalaz/std/AnyValInstances$booleanInstance$disjunction$.class
/Users/jz/.ivy2/cache/org.specs2/scalaz-effect_2.10/jars/scalaz-effect_2.10-7.0.0.jar
scala.MatchError: Z (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)

/Users/jz/.ivy2/cache/org.specs2/specs2-scalaz-core_2.10/jars/specs2-scalaz-core_2.10-7.0.0.jar
scala.MatchError: Z (of class java.lang.Character)
	at scala.tools.nsc.backend.jvm.analysis.BackendUtils$GenericSignatureVisitor$Parser.referenceTypeSignature(BackendUtils.scala:658)

```

## Corrupt class file? ASM bug?

```
/Users/jz/.ivy2/cache/com.ibm.icu/icu4j/jars/icu4j-2.6.1.jar
currentClass = /com/ibm/icu/impl/data/LocaleElements_zh__PINYIN.class
java.lang.ArrayIndexOutOfBoundsException: 48188
	at scala.tools.asm.ClassReader.readClass(ClassReader.java:2461)
	at scala.tools.asm.ClassReader.accept(ClassReader.java:542)
```